Migration of ZK State to 0.13/0.14 does not work

Started some tasks in 0.11 and restarted on master. Got following stack trace:

marathon [2015-10-12 16:19:56,348] ERROR Migration failed! (mesosphere.marathon.state.Migration:97)
marathon java.lang.RuntimeException: while migrating storage to 0.13
marathon 	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:62)
marathon 	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:61)
marathon 	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
marathon 	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:216)
marathon 	at scala.util.Try$.apply(Try.scala:192)
marathon 	at scala.util.Failure.recover(Try.scala:216)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324)
marathon 	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324)
marathon 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
marathon 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
marathon 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
marathon 	at java.lang.Thread.run(Thread.java:745)
marathon Caused by: mesosphere.marathon.StoreCommandFailedException: Could not fetch MarathonTaskState with key: app:app.8f61d32e-70e6-11e5-a216-7a090c3edc79
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$mesosphere$marathon$state$MarathonStore$$exceptionTransform$1.applyOrElse(MarathonStore.scala:75)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$mesosphere$marathon$state$MarathonStore$$exceptionTransform$1.applyOrElse(MarathonStore.scala:74)
marathon 	... 10 more
marathon Caused by: com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.
marathon 	at com.google.protobuf.InvalidProtocolBufferException.invalidEndTag(InvalidProtocolBufferException.java:94)
marathon 	at com.google.protobuf.CodedInputStream.checkLastTagWas(CodedInputStream.java:124)
marathon 	at com.google.protobuf.AbstractParser.parsePartialFrom(AbstractParser.java:143)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:176)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:188)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:193)
marathon 	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:49)
marathon 	at mesosphere.marathon.Protos$MarathonTask.parseFrom(Protos.java:8377)
marathon 	at mesosphere.marathon.state.MarathonTaskState.mergeFromProto(MarathonTaskState.scala:15)
marathon 	at mesosphere.marathon.state.MarathonTaskState.mergeFromProto(MarathonTaskState.scala:9)
marathon 	at mesosphere.marathon.state.MarathonStore.mesosphere$marathon$state$MarathonStore$$stateFromBytes(MarathonStore.scala:79)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1$$anonfun$apply$2.apply(MarathonStore.scala:35)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1$$anonfun$apply$2.apply(MarathonStore.scala:33)
marathon 	at scala.Option.map(Option.scala:146)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1.apply(MarathonStore.scala:33)
marathon 	at mesosphere.marathon.state.MarathonStore$$anonfun$fetch$1.apply(MarathonStore.scala:33)
marathon 	at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
marathon 	at scala.util.Try$.apply(Try.scala:192)
marathon 	at scala.util.Success.map(Try.scala:237)
marathon 	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
marathon 	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
marathon 	... 4 more

--- a/changelog.md
+++ b/changelog.md
@@ -1,3 +1,14 @@
+## Changes from 0.12.0 to 0.13.0
+
+### Breaking Changes
+
+#### Tasks keys and storage format in ZooKeeper
+Marathon tasks are now stored in ZooKeeper using a generic implementation that has been around for
+a while. In order to accomplish this, the keys under which tasks are stored had to be migrated and
+do no longer contain redundant information about the app id. Additionally, the task storage format
+in Zookeeper changed as well. Previous versions of Marathon will **not** be able to read the tasks'
+status once these are migrated. Please backup your ZooKeeper state before migrating to this version. 
+
 ## Changes from 0.11.1 to 0.12.0
 
 ### Overview

--- a/src/main/scala/mesosphere/marathon/state/Migration.scala
+++ b/src/main/scala/mesosphere/marathon/state/Migration.scala
@@ -58,7 +58,7 @@ class Migration @Inject() (
       }
     },
     StorageVersions(0, 13, 0) -> { () =>
-      new MigrationTo0_13(taskRepo).migrateTasks().recover {
+      new MigrationTo0_13(taskRepo, store).migrateTasks().recover {
         case NonFatal(e) => throw new RuntimeException("while migrating storage to 0.13", e)
       }
     }
@@ -286,10 +286,36 @@ class MigrationTo0_11(groupRepository: GroupRepository, appRepository: AppReposi
   }
 }
 
-class MigrationTo0_13(taskRepository: TaskRepository) {
+class MigrationTo0_13(taskRepository: TaskRepository, store: PersistentStore) {
   private[this] val log = LoggerFactory.getLogger(getClass)
 
   val entityStore = taskRepository.store
+
+  // the bytes stored via TaskTracker are incompatible to EntityRepo, so we have to parse them 'manually'
+  def fetchLegacyTask(taskKey: String): Future[Option[MarathonTask]] = {
+    def deserialize(taskKey: String, source: ObjectInputStream): Option[MarathonTask] = {
+      if (source.available > 0) {
+        try {
+          val size = source.readInt
+          val bytes = new Array[Byte](size)
+          source.readFully(bytes)
+          Some(MarathonTask.parseFrom(bytes))
+        }
+        catch {
+          case e: com.google.protobuf.InvalidProtocolBufferException =>
+            None
+        }
+      }
+      else {
+        None
+      }
+    }
+
+    store.load("task:" + taskKey).map(_.flatMap { entity =>
+      val source = new ObjectInputStream(new ByteArrayInputStream(entity.bytes.toArray))
+      deserialize(taskKey, source)
+    })
+  }
 
   def migrateTasks(): Future[Unit] = {
     log.info("Start 0.13 migration")
@@ -299,8 +325,8 @@ class MigrationTo0_13(taskRepository: TaskRepository) {
     // in the key after the prefix implicitly denotes a versioned entry, so this
     // had to be changed, even though tasks are not stored with versions.
     def migrateKey(legacyKey: String): Future[Unit] = {
-      entityStore.fetch(legacyKey).flatMap {
-        case Some(taskState) => taskRepository.store(taskState.task).flatMap { _ =>
+      fetchLegacyTask(legacyKey).flatMap {
+        case Some(task) => taskRepository.store(task).flatMap { _ =>
           entityStore.expunge(legacyKey).map(_ => ())
         }
         case _ => Future.failed[Unit](new RuntimeException(s"Unable to load entity with key = $legacyKey"))

--- a/src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala
+++ b/src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala
@@ -4,14 +4,13 @@ import java.util.UUID
 
 import com.codahale.metrics.MetricRegistry
 import com.fasterxml.uuid.{ EthernetAddress, Generators }
+import mesosphere.FutureTestSupport._
 import mesosphere.marathon.MarathonSpec
 import mesosphere.marathon.Protos.MarathonTask
 import mesosphere.marathon.metrics.Metrics
-import mesosphere.util.state.memory.InMemoryStore
-import org.mockito.Mockito.spy
-import org.scalatest.{ GivenWhenThen, Matchers }
 import mesosphere.marathon.state.PathId.StringPathId
-import mesosphere.FutureTestSupport._
+import mesosphere.util.state.memory.InMemoryStore
+import org.scalatest.{ GivenWhenThen, Matchers }
 
 class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers {
 
@@ -21,8 +20,8 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers 
     val appId = "/test/app1".toRootPath
     val task1 = dummyTask(appId)
     val task2 = dummyTask(appId)
-    f.entityStore.store(f.legacyStoreKey(appId, task1.getId), MarathonTaskState(task1)).futureValue
-    f.entityStore.store(f.legacyStoreKey(appId, task2.getId), MarathonTaskState(task2)).futureValue
+    f.legacyTaskStore.store(appId, task1).futureValue
+    f.legacyTaskStore.store(appId, task2).futureValue
     val names = f.entityStore.names().futureValue
     names should not be empty
     names should contain (appId.safePath + ":" + task1.getId)
@@ -44,8 +43,9 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers 
 
   class Fixture {
     lazy val uuidGenerator = Generators.timeBasedGenerator(EthernetAddress.fromInterface())
-    lazy val state = spy(new InMemoryStore)
+    lazy val state = new InMemoryStore
     lazy val metrics = new Metrics(new MetricRegistry)
+    lazy val legacyTaskStore = new LegacyTaskStore(state)
     lazy val entityStore = new MarathonStore[MarathonTaskState](
       store = state,
       metrics = metrics,
@@ -55,9 +55,42 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers 
       val metrics = new Metrics(new MetricRegistry)
       new TaskRepository(entityStore, metrics)
     }
-    lazy val migration = new MigrationTo0_13(taskRepo)
+    lazy val migration = new MigrationTo0_13(taskRepo, state)
 
     def legacyStoreKey(appId: PathId, taskId: String): String = appId.safePath + ":" + taskId
+  }
+
+}
+
+import java.io._
+
+import mesosphere.util.state.{ PersistentEntity, PersistentStore }
+
+import scala.concurrent.Future
+
+private[state] class LegacyTaskStore(store: PersistentStore) {
+
+  val PREFIX = "task:"
+  val ID_DELIMITER = ":"
+
+  private[this] def getKey(appId: PathId, taskId: String): String = {
+    PREFIX + appId.safePath + ID_DELIMITER + taskId
+  }
+
+  private[this] def serialize(task: MarathonTask, sink: ObjectOutputStream): Unit = {
+    val size = task.getSerializedSize
+    sink.writeInt(size)
+    sink.write(task.toByteArray)
+    sink.flush()
+  }
+
+  def store(appId: PathId, task: MarathonTask): Future[PersistentEntity] = {
+    val byteStream = new ByteArrayOutputStream()
+    val output = new ObjectOutputStream(byteStream)
+    serialize(task, output)
+    val bytes = byteStream.toByteArray
+    val key: String = getKey(appId, task.getId)
+    store.create(key, bytes)
   }
 
 }

